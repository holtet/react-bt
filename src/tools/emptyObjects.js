export const malt = {
  comments: '',
  grams: 0,
  boughtDate: new Date(),
  maltDefinition: { name: '', description: '', manufacturer: '', country: '' },
};
export const maltType = {
  name: '',
  country: '',
  manufacturer: '',
  description: '',
};
export const yeast = {
  grams: 0,
  packages: 0,
  comments: '',
  boughtDate: new Date(),
  yeastType: { name: '', format: '' },
};
export const equipment = {
  boughtDate: new Date(),
  comments: '',
  amount: 0,
  equipmentType: {
    name: '',
    description: '',
  },
};
export const hop = {
  grams: 0,
  year: 2017,
  alpha: 0.0,
  comments: '',
  boughtDate: new Date(),
  hopType: { name: '', format: '' },
};
export const hopType = {
  name: '',
  manufacturer: '',
  country: '',
  description: '',
};
