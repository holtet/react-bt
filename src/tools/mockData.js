const yeastTypes = [
  {
    name: 'Saf5',
    manufacturer: 'Safale',
    country: 'England',
    format: 'DRY',
    _links: {
      self: {
        href: 'http://localhost:8080/yeastTypes/14',
      },
      yeastType: {
        href: 'http://localhost:8080/yeastTypes/14',
      },
    },
  },
  {
    name: 'WLP400',
    manufacturer: 'WyEast',
    country: 'Belgium',
    format: 'LIQUID',
    _links: {
      self: {
        href: 'http://localhost:8080/yeastTypes/15',
      },
      yeastType: {
        href: 'http://localhost:8080/yeastTypes/15',
      },
    },
  },
];
const hopTypes = [
  {
    name: 'Cascade',
    manufacturer: 'US Hop Co',
    country: 'USA',
    format: 'PELLETS',
    _links: {
      self: {
        href: 'http://localhost:8080/hopTypes/10',
      },
      hopType: {
        href: 'http://localhost:8080/hopTypes/10',
      },
    },
  },
  {
    name: 'hallertau',
    manufacturer: 'Hops GMBH',
    country: 'Germany',
    format: 'PELLETS',
    _links: {
      self: {
        href: 'http://localhost:8080/hopTypes/11',
      },
      hopType: {
        href: 'http://localhost:8080/hopTypes/11',
      },
    },
  },
];
const hops = [
  {
    year: 2017,
    alpha: 3.7,
    grams: 90,
    boughtDate: '2020-05-29',
    comments: '',
    openedDate: null,
    hopType: {
      name: 'Cascade',
      manufacturer: 'US Hop Co',
      country: 'USA',
      format: 'PELLETS',
    },
    type: {
      name: 'Cascade',
      manufacturer: 'US Hop Co',
      country: 'USA',
      format: 'PELLETS',
    },
    _links: {
      self: {
        href: 'http://localhost:8080/hops/12',
      },
      hop: {
        href: 'http://localhost:8080/hops/12{?projection}',
        templated: true,
      },
      type: {
        href: 'http://localhost:8080/hops/12/type',
      },
    },
  },
  {
    year: 2018,
    alpha: 5.3,
    grams: 15,
    boughtDate: '2020-05-29',
    comments: '',
    openedDate: null,
    hopType: {
      name: 'hallertau',
      manufacturer: 'Hops GMBH',
      country: 'Germany',
      format: 'PELLETS',
    },
    _links: {
      self: {
        href: 'http://localhost:8080/hops/13',
      },
      hop: {
        href: 'http://localhost:8080/hops/13{?projection}',
        templated: true,
      },
      type: {
        href: 'http://localhost:8080/hops/13/type',
      },
    },
  },
];
const apiHops = { _embedded: { hops: hops } };
const newHop = { id: null, comments: '' };
module.exports = {
  hopTypes,
  hops,
  apiHops,
  newHop,
  yeastTypes,
};
