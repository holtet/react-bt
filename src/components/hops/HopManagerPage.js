import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import itemActions from '../../redux/actions/hopActions';
import PropTypes from 'prop-types';
import Spinner from '../common/Spinner';
import itemTypeActions from '../../redux/actions/hopTypeActions';
import { toast } from 'react-toastify';
import ItemForm from '../common/ItemForm';
import { hop as hopTemplate } from '../../tools/emptyObjects';
import { itemWrapper } from '../common/ItemWrapper';

function capitalizeFirst(string) {
  return string.charAt(0).toUpperCase() + string.slice(1).toLowerCase();
}
function HopManagerPage({
  itemTypes,
  items,
  loadItems,
  loadItemTypes,
  saveItem,
  history,
  ...props
}) {
  const [item, setItem] = useState(props.item);
  const [errors, setErrors] = useState({});
  const [saving, setSaving] = useState(false);

  useEffect(() => {
    if (items.length === 0) {
      console.log('Loading hops');
      loadItems().catch((error) => {
        alert('loading hops failed ' + error);
      });
    } else {
      setItem({ ...props.item });
    }
    if (itemTypes.length === 0) {
      console.log('Loading hopTypes');
      loadItemTypes().catch((error) => {
        alert('loading hopDefs failed ' + error);
      });
    }
  }, [props.item, props.itemTypes]); //TODO: props.hopTypes => hopTypes??

  function formIsValid() {
    const { grams, hopType, newItemType } = item;
    console.log('hopType' + JSON.stringify(hopType));
    const errors = {};
    if (isNaN(grams)) errors.grams = 'Not a valid number';
    if (!(hopType.name || newItemType)) errors.type = 'Hop type is required';
    setErrors(errors);
    // Form is valid if the errors object still has no properties
    return Object.keys(errors).length === 0;
  }

  function handleSave(event) {
    console.log('Handle save');
    event.preventDefault();
    if (!formIsValid()) return;
    setSaving(true);
    saveItem(item)
      .then(() => {
        toast.success('Hop saved.');
        history.push('/hops');
      })
      .catch((error) => {
        setSaving(false);
        setErrors({ onSave: error.message });
      });
  }

  function handleChange(event) {
    const { name, value } = event.target;
    console.log('Handle change ' + name + ' = ' + value);

    if (name == 'type' && value != '') {
      setItem((prevItem) => ({
        ...prevItem,
        newItemType: value,
        hopType: itemTypes.find((h) => h._links.self.href == value),
      }));
    } else {
      setItem((prevItem) => ({
        ...prevItem,
        [name]: value,
      }));
    }
  }

  const fields = [
    { name: 'Grams', field: 'grams', type: 'text' },
    { name: 'Year', field: 'year', type: 'text' },
    { name: 'Alpha', field: 'alpha', type: 'text' },
    { name: 'Comments', field: 'comments', type: 'textArea' },
    {
      name: 'Bought Date',
      field: 'boughtDate',
      type: 'date',
      transform: (a) => new Date(a),
    },
    {
      name: 'Opened Date',
      field: 'openedDate',
      type: 'date',
      transform: (a) => (a == undefined ? undefined : new Date(a)),
    },
  ];

  const typeFields = [
    { name: 'Name', field: 'name', link: true },
    { name: 'Country', field: 'country' },
    { name: 'Format', field: 'format', transform: (a) => capitalizeFirst(a) },
    { name: 'Manufacturer', field: 'manufacturer' },
  ];
  return itemTypes.length == 0 ? (
    <Spinner />
  ) : (
    <ItemForm
      itemTypes={itemTypes} //.filter((it) => it.name != item.hopType.name)}
      item={item}
      itemType={item.hopType}
      errors={errors}
      onChange={handleChange}
      fields={fields}
      typeFields={typeFields}
      typeOptionFormatter={(type) =>
        type.name + ' | ' + capitalizeFirst(type.format)
      }
      onSave={handleSave}
      saving={saving}
    />
  );
}

HopManagerPage.propTypes = {
  loadItems: PropTypes.func.isRequired,
  loadItemTypes: PropTypes.func.isRequired,
  saveItem: PropTypes.func.isRequired,
  itemTypes: PropTypes.array.isRequired,
  item: PropTypes.object.isRequired,
  items: PropTypes.array.isRequired,
  history: PropTypes.object.isRequired,
};

function mapStateToProps(state, ownProps) {
  const index = ownProps.match.params.index;
  const item =
    index && state.items.hops.length > 0
      ? state.items.hops[index]
      : hopTemplate;
  // console.log('Index: ' + index);
  // console.log('Hop year: ' + hop.year);
  // let hopType = { name: '' };
  // if (hop._links != undefined) {
  //   hopType = state.getHopType(hop._links.hopType.href);
  //   console.log('Fetched hopDef' + hopType);
  // }
  return {
    item,
    items: state.items.hops,
    itemTypes: state.items.hopTypes,
  };
}

const mapDispatchToProps = {
  loadItems: itemActions.loadAll,
  saveItem: itemActions.saveItem,
  loadItemTypes: itemTypeActions.loadAll,
};

export default itemWrapper(
  connect(mapStateToProps, mapDispatchToProps)(HopManagerPage)
);
