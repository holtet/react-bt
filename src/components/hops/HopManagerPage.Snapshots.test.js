import React from 'react';
import HopManagerPage from './HopManagerPage';
import renderer from 'react-test-renderer';
import { hops, hopTypes } from '../../tools/mockData';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
const mockStore = configureMockStore();
const store = mockStore({
  hopTypes: hopTypes,
  hops: hops,
  items: { hopTypes: hopTypes, hops: hops },
});

it("sets submit button label 'Saving...' when saving is true", () => {
  console.log(JSON.stringify(hops[0]));

  const tree = renderer.create(
    <Provider store={store}>
      <HopManagerPage
        item={hops[0]}
        itemTypes={hopTypes}
        itemType={hopTypes[0]}
        onSave={jest.fn()}
        onChange={jest.fn()}
        saving
        match={{ params: { index: 0 } }}
        history={{}}
      />
    </Provider>
  );

  expect(tree).toMatchSnapshot();
});
