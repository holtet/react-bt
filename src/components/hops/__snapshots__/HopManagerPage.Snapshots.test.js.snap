// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets submit button label 'Saving...' when saving is true 1`] = `
<div>
  <div
    className="overlay"
    style={
      Object {
        "width": 0,
      }
    }
  >
    <div
      className="sidenav-container"
    >
      <div
        className="text-center"
      >
        <h2>
          Hop
           changelog
        </h2>
      </div>
      <a
        className="closebtn"
        onClick={[Function]}
      >
        Ã—
      </a>
      <div
        className="list-group"
      >
        <table
          className="table table-dark"
          style={
            Object {
              "width": "95%",
            }
          }
        >
          <thead>
            <tr>
              <td>
                Time and date
              </td>
              <td>
                Field
              </td>
              <td>
                Changed from
              </td>
              <td>
                To
              </td>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </div>
  <div
    className="card"
    style={
      Object {
        "width": "18rem",
      }
    }
  >
    <div
      className="card-header"
    >
      Hop Type
    </div>
    <div
      className="card-body"
    >
      <table
        className="table table-sm"
      >
        <tbody>
          <tr>
            <td>
              Name
            </td>
            <td>
              
            </td>
          </tr>
          <tr>
            <td>
              Country
            </td>
            <td />
          </tr>
          <tr>
            <td>
              Format
            </td>
            <td>
              
            </td>
          </tr>
          <tr>
            <td>
              Manufacturer
            </td>
            <td />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <form
    onSubmit={[Function]}
    style={
      Object {
        "width": "18rem",
      }
    }
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="grams"
      >
        grams
      </label>
      <div
        className="field"
      >
        <input
          className="form-control"
          name="grams"
          onChange={[Function]}
          type="text"
          value="0"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="year"
      >
        year
      </label>
      <div
        className="field"
      >
        <input
          className="form-control"
          name="year"
          onChange={[Function]}
          type="text"
          value="2017"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="alpha"
      >
        alpha
      </label>
      <div
        className="field"
      >
        <input
          className="form-control"
          name="alpha"
          onChange={[Function]}
          type="text"
          value="0"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="comments"
      >
        comments
      </label>
      <div
        className="field"
      >
        <textarea
          className="form-control"
          name="comments"
          onChange={[Function]}
          value=""
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="boughtDate"
      >
        boughtDate
      </label>
      <div
        className="field"
      >
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              className="form-control"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              readOnly={false}
              type="text"
              value="02.06.2020"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="openedDate"
      >
        openedDate
      </label>
      <div
        className="field"
      >
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              className="form-control"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              readOnly={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="hopType"
      >
        hopType
      </label>
      <div
        className="field"
      >
        <select
          className="form-control"
          name="hopType"
          onChange={[Function]}
        >
          <option
            value=""
          >
             | 
          </option>
          <option
            value="http://localhost:8080/hopTypes/10"
          >
            Cascade | Pellets
          </option>
          <option
            value="http://localhost:8080/hopTypes/11"
          >
            hallertau | Pellets
          </option>
        </select>
      </div>
    </div>
    <button
      className="btn btn-primary"
      disabled={false}
      type="submit"
    >
      Save
    </button>
  </form>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Log
  </button>
</div>
`;
